source/auth/tests/connection-string.json

```diff
@@ -1,279 +1,344 @@
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "user",
+        "password": "password",
+        "db": "admin"
+      },
+      "options": {
+        "authmechanism": "MONGODB-CR"
+      }
+    },
+    {
+      "description": "should throw an exception when an authMechanism (MONGODB-CR) is specified with no username",
+      "uri": "mongodb://localhost/?authMechanism=MONGODB-CR",
+      "hosts": null,
+      "valid": false,
+      "warning": true,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should recognize the mechanism (MONGODB-X509)",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry",
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should recognize the mechanism where authSource $external (MONGODB-X509)",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/foo?authMechanism=MONGODB-X509&authSource=$external",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": "CN=myName,OU=myOrgUnit,O=myOrg,L=myLocality,ST=myState,C=myCountry",
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should throw an exception when an authMechanism (MONGODB-X509) has an authSource other than $external",
+      "uri": "mongodb://CN%3DmyName%2COU%3DmyOrgUnit%2CO%3DmyOrg%2CL%3DmyLocality%2CST%3DmyState%2CC%3DmyCountry@localhost/foo?authMechanism=MONGODB-X509&authSource=bar",
+      "hosts": null,
+      "valid": false,
+      "warning": true,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should recognize the mechanism (MONGODB-X509) with no username",
+      "uri": "mongodb://localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": true,
+      "warning": false,
+      "auth": {
+        "username": null,
+        "password": null,
+        "db": "$external"
+      },
+      "options": {
+        "authmechanism": "MONGODB-X509"
+      }
+    },
+    {
+      "description": "should not support passwords with the mechanism (MONGODB-X509)",
+      "uri": "mongodb://user:password@localhost/?authMechanism=MONGODB-X509",
+      "hosts": null,
+      "valid": false,
+      "warning": false,
+      "auth": null,
+      "options": null
+    },
+    {
+      "description": "should throw an exception if provided with an invalid authSource (MONGODB-X509)",
```

Comment:
You're already testing this two tests up.

Reply:
Good catch - removed